exercises/binary-search-tree/canonical-data.json

```diff
@@ -0,0 +1,68 @@
+{
+    "exercise": "binary-search-tree",
+    "version": "1.0.0",
+    "comments": [
+        "For simplicity only positive integer is considered as valid data,",
+        "null should be replaced by the respective language equivalent or -1",
+        "if the language does not have any.",
+
+        "The purpose of ths exercise is to create a data structure which",
+        "makes sure the data is always sorted, hence the exercise should",
+        "not make use of any sort() functions or sorting algorithms."
+    ],
+    "cases": [{
+            "description": "insert invalid/null data into the tree",
+            "property": "insert",
+            "value": null,
+            "expected": false
+        },
+        {
+            "description": "insert a number into the tree",
+            "property": "insert",
+            "value": 6,
+            "expected": true
+        },
+        {
+            "description": "insert a number that already exist",
+            "property": "insert",
+            "value": 6,
+            "expected": false
+        },
+        {
+            "description": "search invalid/null data",
+            "property": "search",
+            "value": null,
+            "expected": false
+        },
+        {
+            "description": "search a number that exists",
+            "property": "search",
+            "value": 6,
+            "expected": true
+        },
+        {
+            "description": "search a number that does not exist",
+            "property": "search",
+            "value": 11,
+            "expected": false
+        },
+        {
+            "description": "traverse the data in ascending order",
+            "property": "ascendingOrder",
+            "comments": [
+                "Assuming the data is entered in this order :",
+                "6 -> 1 -> 3 -> 2 -> 5"
```

Comment:
it looks like these values should be an array of numbers, rather than forcing people to parse this string, splitting by `->` to find the numbers, etc

Reply:
I thought about it but i wasn't sure if i should add a new key called "state" which represents the data entered in the tree at that case in that specific order. This new key "state" will also prove useful in previous cases where its not clear what data the tree currently has.

Reply:
I think there has to be such a key. You could call it `previously_inserted_values` instead of `state` if needing a more descriptive name.

Reply:
Um. maybe `current_tree_data` because then i can also have this key in all cases and make all cases consistent.

Reply:
I'm not even sure how to read this...

> 6, then 1, then 3, then...

or is it

> 6 after, 1 after 3, after...

or

> 6 has child 1, has child 3, has child (but which child then?)

Some implementations ask for implementing a function that creates a BST from a list, growing the tree from the first item to the last.

Roughly like this:

```hs
foldl Leaf (\ x t -> insert t x) list
```

In the introductionary comment we could explain that a property `"value"` could be used to create a singleton tree, or inserting a single value to a given tree, while `"values"` or `"tree"` implies to create a tree from a list using this function.